### PR TITLE
[MaterializeBlockPointer] Fix wrong per-warp base in 1D strided load reshape

### LINKED ISSUE #7918: 1D-strided -> 2D-block-load rewrite produces incorrect results on BMG
## Problem

`MaterializeBlockPointer`'s rewrite of a 1D strided load into a 2D block load returns wrong data on Intel Arc B580 (BMG). The path is enabled by default, so affected kernels silently produce incorrect results.

## Reproduction

```python
import torch, triton, triton.language as tl

BLOCK, ROW, STRIDE = 4096, 32, 96

@triton.jit
def k(in_ptr, out_ptr, BLOCK: tl.constexpr, ROW: tl.constexpr, STRIDE: tl.constexpr):
    off = tl.arange(0, BLOCK)
    src = (off % ROW) + (off // ROW) * STRIDE
    tl.store(out_ptr + off, tl.load(in_ptr + src))

dev = 'xpu'
N = (BLOCK // ROW) * STRIDE
inp = torch.arange(N, dtype=torch.float32, device=dev).to(torch.float16)
out = torch.empty((BLOCK,), dtype=torch.float16, device=dev)
k[(1,)](inp, out, BLOCK, ROW, STRIDE)
torch.xpu.synchronize()

idx = torch.arange(BLOCK, device=dev)
ref = inp[(idx % ROW) + (idx // ROW) * STRIDE]
bad = (out != ref)
print('mismatches:', bad.sum().item(), 'of', BLOCK)
w = torch.nonzero(bad).flatten()
if w.numel():
    i = w[0].item()
    print(f'first bad at {i}: got {out[i].item()} want {ref[i].item()}')
```

```
$ python repro.py
mismatches: 3072 of 4096
first bad at 1024: got 768.0 want 3072.0
```

Device: Intel(R) Arc(TM) B580 Graphics.

### PR DESCRIPTION
**Pull request by etiotto**
The 1D->2D load rewrite reshaped the pointer and mask with
`allow_reorder efficient_layout`. That reshape only changes the type, it
does not move data, so the pointer tensor claimed the layout the 2D block
load expects while the registers still held pointers in the original 1D
order.

The lowering reads each warp's base address out of that tensor, so every
warp but the first started at the wrong row. With BLOCK=4096, W=32, S=96
and num_warps=4, warp w read row 8*w instead of 32*w and 3072 of 4096
elements came back wrong.

Reshape the pointer and mask without reordering, then convert them to the
load layout with ttg::ConvertLayoutOp. The store path already does this,
which is why stores were never affected. The lowering is unchanged.

The convert costs nothing: RemoveLayoutConversions pushes the layout back
into the addptr chain and folds it away. The generated LLVM IR keeps the
same 8192 B of shared memory, one barrier and one LSC2DBlockRead as before.

Update the three lit tests that matched the old pointer reshape.

Fixes #7918

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUTransforms/MaterializeBlockPointer.cpp [hunk 1]
```diff
@@ -880,31 +880,35 @@ struct TritonIntelGPUMaterializeBlockPointerPass
     if (!consumerEnc)
```
Comment (chengjunlu) [inline third_party/intel/lib/TritonIntelGPUTransforms/MaterializeBlockPointer.cpp:880]: The consumer encoding is not used since the ReshapeOp layout is automatically deduced by the op itself.

Can we remove it?

File: test/TritonIntelGPU/materialize-block-pointer-1d-reshape.mlir [hunk 2]
```diff
@@ -158,11 +158,18 @@ module attributes {"ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 1 : i32, "ttg.thr
...
+  // COM: The pointer and mask must be reshaped without reordering and then
+  // COM: physically converted into the load encoding. A relabelling reshape
+  // COM: (allow_reorder efficient_layout) moves no data, so the operands would
+  // COM: claim the load encoding while still holding 1D-ordered values.
   // CHECK-LABEL: tt.func @test_1d_strided_load
-  // CHECK: tt.reshape %{{.*}} allow_reorder efficient_layout
-  // CHECK: tt.load %{{.*}} {ttig.block_io = "row_major", ttig.block_io_stride = 96 : i64}
-  // CHECK: ttg.convert_layout
-  // CHECK: tt.reshape %{{.*}} efficient_layout
+  // CHECK: [[PTR2D:%[0-9]+]] = tt.reshape %{{.*}} : tensor<1024x!tt.ptr<f16>, {{.*}}> -> tensor<32x32x!tt.ptr<f16>, [[CONSENC:#[a-z0-9_]+]]>
```
[the remark this answers is not in the document]
Comment (etiotto) [reply test/TritonIntelGPU/materialize-block-pointer-1d-reshape.mlir:166]: You're right about the fixture: at 1024 elements with 4 warps `reps1D = 1024/(8*32*4) = 1`, so the per-warp height is 8 and the 1D layout also places warp `w` at row `8*w`. The two encodings coincide there, and this `.mlir` test is intentionally IR-shape-only — it pins the rewrite's structure, which is all `triton-opt` + FileCheck can check.

The executing regression you're asking for is already in this PR, in `python/test/unit/intel/test_1d_reshape_store_correctness.py`: `test_1d_reshape_strided_load_multi_warp_base`. Its first parametrization is exactly `XBLOCK=4096, W=32, S=96, num_warps=4` (the #7918 shape, `mismatches: 3072 -> 0`), plus `S=128`/`S=192`, `num_warps=2` (`XBLOCK=2048`), and `num_warps=8` (per-warp height 16, below the HW maximum). Every config has `reps1D >= 2` and `num_warps >= 2`, i.e. the regime where a wrong per-warp base is observable. It also asserts `_has_block_load(...)` on the LLIR so the test cannot silently pass if the 1D->2D rewrite stops firing.

No code change needed here; resolving.

File: third_party/intel/lib/TritonIntelGPUTransforms/MaterializeBlockPointer.cpp (hunk 1 above)
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUTransforms/MaterializeBlockPointer.cpp:880]: You're right that it's redundant for the reshapes themselves. Now that the pointer and mask use the shape-only `ReshapeOp` builder, they deduce the encoding via the same `inferReshapeOpEncoding(..., allowReorder=false)` call that `derive2DConsumerEncoding` makes, so it's the identical attribute.

It's still used further down though, as the target type of the `ConvertLayoutOp` on the load result (`loadEnc` -> natural 2D):

```cpp
auto consumerResultTy = RankedTensorType::get(newShape, pointeeTy, *consumerEnc);
auto converted = ttg::ConvertLayoutOp::create(builder, loc, consumerResultTy, newLoad);
```

That's the convert `materialize-block-pointer-1d-reshape-anchor.mlir` is guarding, so I'd rather keep it.

File: test/TritonIntelGPU/materialize-block-pointer-1d-reshape-anchor.mlir [hunk 3]
```diff
@@ -1,15 +1,22 @@
 // RUN: triton-opt %s -split-input-file -tritonintelgpu-materialize-block-pointer -tritonintelgpu-remove-layout-conversions | FileCheck %s
 
-// COM: Regression test: verify that RemoveLayoutConversions does NOT remove
-// COM: the ConvertLayoutOp inserted by reshape1DStridedLoad.  The load encoding
-// COM: matches HW delivery order and must be anchored; without the anchor fix
-// COM: in isExpensiveLoadOrStore the ConvertLayoutOp is eliminated and the load
+// COM: Regression test for how RemoveLayoutConversions treats the two
+// COM: ConvertLayoutOps that reshape1DStridedLoad inserts.
```
[the remark this answers is not in the document]
Comment (etiotto) [reply test/TritonIntelGPU/materialize-block-pointer-1d-reshape-anchor.mlir]: Correct, thanks — this fixture's load is masked, so the rewrite converts the mask operand too: pointer, mask and result, not two.

Fixed in 53a0facf9. The comment now describes them by role rather than by count (the mask convert only exists when the load is masked) and says explicitly that the pointer and mask ones are expected to fold away, which is what the existing `CHECK-NOT: ttg.convert_layout` between the pointer reshape and the load already pins.

Comment text only — no CHECK lines and no code changed.

**Review by chengjunlu — APPROVED** [review]
